firestore-next/test.firestore.js

```diff
@@ -1323,3 +1323,1294 @@ describe("firestore", () => {
+            field("release_date").descending(), field("author").ascending()
+          );
+        // [END sort_comparison]
+        console.log(q);
+        console.log(pipeline);
+    }
+
+    async function functions() {
+        // [START functions_example]
+        let results;
+
+        // Type 1: Scalar (for use in non-aggregation stages)
+        // Example: Return the min store price for each book.
+        results = await execute(db.pipeline().collection("books")
+          .select(
+            field("current").logicalMinimum(["updated"]).as("price_min")
+          )
+        );
+
+        // Type 2: Aggregation (for use in aggregate stages)
+        // Example: Return the min price of all books.
+        results = await execute(db.pipeline().collection("books")
+          .aggregate(field("price").minimum().as("min_price"))
+        );
+        // [END functions_example]
+        console.log(results);
+    }
+
+    async function creatingIndexes() {
+        // [START query_example]
+        const results = await execute(db.pipeline()
+          .collection("books")
+          .where(field("published").lessThan(1900))
+          .where(field("genre").equal("Science Fiction"))
+          .where(field("rating").greaterThan(4.3))
+          .sort(field("published").descending())
+        );
+        // [END query_example]
+        console.log(results);
+    }
+
+    async function sparseIndexes() {
+        // [START sparse_index_example]
+        const results = await execute(db.pipeline()
+          .collection("books")
+          .where(field("category").like("%fantasy%"))
+        );
+        // [END sparse_index_example]
+        console.log(results);
+    }
+
+    async function sparseIndexes2() {
+        // [START sparse_index_example_2]
+        const results = await execute(db.pipeline()
+          .collection("books")
+          .sort(field("release_date").ascending())
+        );
+        // [END sparse_index_example_2]
+        console.log(results);
+    }
+
+    async function coveredQuery() {
+        // [START covered_query]
+        const results = await execute(db.pipeline()
+          .collection("books")
+          .where(field("category").like("%fantasy%"))
+          .where(field("title").exists())
+          .where(field("author").exists())
+          .select(field("title"), field("author"))
+        );
+        // [END covered_query]
+        console.log(results);
+    }
+
+    function pagination() {
+        // [START pagination_not_supported_preview]
+        // Existing pagination via `startAt()`
+        const q = // db.collection("cities").orderBy("population").startAt(1000000);
+          query(collection(db, "cities"), orderBy("population"), startAt(1000000));
+
+        // Private preview workaround using pipelines
+        const pipeline = db.pipeline()
+          .collection("cities")
+          .where(field("population").greaterThanOrEqual(1000000))
+          .sort(field("population").descending());
```

Comment:
I have two approaches to pagination that handle collisions. 

https://github.com/firebase/firebase-js-sdk/blob/5206828ce60741227a9ac8f7ab358502c04e9ab5/packages/firestore/test/integration/api/pipeline.test.ts#L4196

Reply:
Done.